plugin.js

```diff
@@ -66,10 +58,42 @@ function onReqHandlerWrapper (fastify, hook) {
       if (cookieHeader) {
         fastifyReq.cookies = fastify.parseCookie(cookieHeader)
       }
+      fastifyRes[kReplySetCookies] = new Map()
       done()
     }
 }
 
+function fastifyCookieOnSendHandler (fastifyReq, fastifyRes, payload, done) {
+  if (fastifyRes[kReplySetCookies].size) {
+    let setCookie = fastifyRes.getHeader('Set-Cookie')
+
+    /* istanbul ignore else */
+    if (!setCookie) {
+      if (fastifyRes[kReplySetCookies].size === 1) {
+        for (const c of fastifyRes[kReplySetCookies].values()) {
+          setCookie = cookie.serialize(c.name, c.value, c.opts)
+        }
+
+        fastifyRes.header('Set-Cookie', setCookie)
+        return done()
```

Comment:
we could avoid the for loop

```suggestion
        const c = fastifyRes[kReplySetCookies].values()[0]
        fastifyRes.header('Set-Cookie', cookie.serialize(c.name, c.value, c.opts))
        return done()
```

Reply:
Got this with:
```js
      if (fastifyRes[kReplySetCookies].size === 1) {
        const c = fastifyRes[kReplySetCookies].values().next().value

        fastifyRes.header('Set-Cookie', cookie.serialize(c.name, c.value, c.opts))

        return done()
      }
```


```shell
100 connections


┌─────────┬──────┬──────┬───────┬──────┬─────────┬─────────┬────────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%  │ Avg     │ Stdev   │ Max    │
├─────────┼──────┼──────┼───────┼──────┼─────────┼─────────┼────────┤
│ Latency │ 1 ms │ 1 ms │ 2 ms  │ 2 ms │ 1.04 ms │ 0.77 ms │ 133 ms │
└─────────┴──────┴──────┴───────┴──────┴─────────┴─────────┴────────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬──────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg      │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼──────────┼─────────┼─────────┤
│ Req/Sec   │ 73471   │ 73471   │ 81535   │ 82495   │ 80759.28 │ 2470.63 │ 73462   │
├───────────┼─────────┼─────────┼─────────┼─────────┼──────────┼─────────┼─────────┤
│ Bytes/Sec │ 15.9 MB │ 15.9 MB │ 17.7 MB │ 17.9 MB │ 17.5 MB  │ 537 kB  │ 15.9 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴──────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.
# of samples: 11

888k requests in 11.01s, 193 MB read
```

Reply:
Ok, at this point, they are trading wins

Should I use this:
```js
      if (fastifyRes[kReplySetCookies].size === 1) {
        const c = fastifyRes[kReplySetCookies].values().next().value

        fastifyRes.header('Set-Cookie', cookie.serialize(c.name, c.value, c.opts))

        return done()
      }
```

or this:
```js
      if (fastifyRes[kReplySetCookies].size === 1) {
        for (const c of fastifyRes[kReplySetCookies].values()) {
          fastifyRes.header('Set-Cookie', cookie.serialize(c.name, c.value, c.opts))
        }

        return done()
      }
```

The first one has an extra method call and returns an object, so we have to call .value on it, `for` just gets the value

Reply:
Use the one you think is the best one.

Reply:
How is the perf of the 5 cookies?

Reply:
Alright, will see!

For 5 cookies, here are the very first tests I ran on both (while my Mac is still cold :))

master:
```shell
10 connections


┌─────────┬──────┬──────┬───────┬──────┬─────────┬─────────┬───────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%  │ Avg     │ Stdev   │ Max   │
├─────────┼──────┼──────┼───────┼──────┼─────────┼─────────┼───────┤
│ Latency │ 0 ms │ 0 ms │ 0 ms  │ 0 ms │ 0.01 ms │ 0.06 ms │ 13 ms │
└─────────┴──────┴──────┴───────┴──────┴─────────┴─────────┴───────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg     │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Req/Sec   │ 54527   │ 54527   │ 66239   │ 66623   │ 64790.4 │ 3485.01 │ 54496   │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Bytes/Sec │ 18.4 MB │ 18.4 MB │ 22.3 MB │ 22.5 MB │ 21.8 MB │ 1.18 MB │ 18.4 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.
# of samples: 10

648k requests in 10.01s, 218 MB read
```

PR:
```shell
10 connections


┌─────────┬──────┬──────┬───────┬──────┬─────────┬─────────┬───────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%  │ Avg     │ Stdev   │ Max   │
├─────────┼──────┼──────┼───────┼──────┼─────────┼─────────┼───────┤
│ Latency │ 0 ms │ 0 ms │ 0 ms  │ 0 ms │ 0.01 ms │ 0.05 ms │ 13 ms │
└─────────┴──────┴──────┴───────┴──────┴─────────┴─────────┴───────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬──────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg      │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼──────────┼─────────┼─────────┤
│ Req/Sec   │ 63935   │ 63935   │ 69823   │ 70143   │ 69240.73 │ 1700.63 │ 63928   │
├───────────┼─────────┼─────────┼─────────┼─────────┼──────────┼─────────┼─────────┤
│ Bytes/Sec │ 20.2 MB │ 20.2 MB │ 22.1 MB │ 22.2 MB │ 21.9 MB  │ 539 kB  │ 20.2 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴──────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.
# of samples: 11

762k requests in 11.01s, 241 MB read
```

Reply:
The numbers budge, but `for...of` is consistently (slightly) faster on my machine

Here is why I think that is: `.values` creates an iterator object but calling next on it creates yet another object while `for...of` just gets the value. Or maybe `for...of` causes an optimization since it's language syntax

Anyway I think we're good